// Copyright (c) 2013, Cloudera, inc.
package kudu.log;

import "consensus/consensus.proto";
import "server/metadata.proto";

// ===========================================================================
//  Log Entries - Log specific messages + single node messages
// ===========================================================================

// Types of log entries.
enum LogEntryTypePB {
  OPERATION = 1;
  // TODO remove this (along with the parent enum) when we change the
  // log GC not to depend on the tablet metadata. Moreover consider
  // making the Log store OperationPB directly, as LogEntryPB might
  // become obsolete.
  TABLET_METADATA = 2;
};

// An entry in the WAL/state machine log.
message LogEntryPB {
  required LogEntryTypePB type = 1;
  optional consensus.OperationPB operation = 2;
  optional metadata.TabletSuperBlockPB tablet_meta = 3;
}

// A header for a log segment.
message LogSegmentHeaderPB {
  // Log format major version.
  required uint32 major_version = 1;

  // Log format minor version.
  required uint32 minor_version = 2;

  // The term/index when the segment was started.
  // TODO: rename to initial_opid.
  required kudu.consensus.OpId initial_id = 3;

  // The tablet metadata at the time the segment was created.
  required metadata.TabletSuperBlockPB tablet_meta = 4;
}
